### PR TITLE
net: add tos and set_tos methods to TCP and UDP sockets

### LINKED ISSUE #3545: TCPStream/UDPSocket tos/set_tos methods
**Is your feature request related to a problem? Please describe.**
Some network applications rely on proper marking of egress packets using ToS/DSCP fields. Routing protocols,
management traffic, VoIP applications and so on. 

ToS can be set on socket level using `setsockopt` call, using `IPPROTO_IP` level and `IP_TOS` option.
TCPStream and UDPSocket already have functions:

*  `ttl`
* `set_ttl`

for getting an setting TTL field. TTL is manipulated via the same `setsockopt`/`getsockopt` calls.

**Describe the solution you'd like**
Adding methods

* `tos`
* `set_tos`

for getting socket's ToS settings will be good addition for a large spectre of network applications.

**Describe alternatives you've considered**
Direct usage of setsockopt on raw socket handle still possible, and can be used as workaround.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by taiki-e**
Closes #3545

cc #3082

The first commit is from #4361.

Refs:
- [socket2::Socket::tos](https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.tos)
- [socket2::Socket::set_tos](https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.set_tos)

### CODE REVIEW

File: tokio/src/net/tcp/socket.rs [hunk 1]
```diff
@@ -378,7 +378,68 @@ impl TcpSocket {
...
+        doc(cfg(all(
+            unix,
+            not(any(
+                target_os = "fuchsia",
+                target_os = "redox",
+                target_os = "solaris",
+                target_os = "illumos",
+            ))
+        )))
+    )]
```
Comment (Darksonn) [inline tokio/src/net/tcp/socket.rs:410]: One of these includes a `all(unix, ...)` and the other doesn't. Why?

Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs:410]: Good catch. I forgot to remove `all(unix, ...)` when I copied the doc cfg from `TcpSocket::reuseport`.

File: tokio/src/net/udp.rs [hunk 2]
```diff
@@ -1503,6 +1503,65 @@ impl UdpSocket {
...
+            target_os = "illumos",
+        ))))
+    )]
+    pub fn set_tos(&self, tos: u32) -> io::Result<()> {
+        self.to_socket().set_tos(tos)
+    }
+
+    fn to_socket(&self) -> socket2::SockRef<'_> {
+        socket2::SockRef::from(self)
+    }
```
Comment (Darksonn) [inline tokio/src/net/udp.rs:1563]: Like I commented on the other PR, it is more idiomatic to call it `as_socket`.
```suggestion
    fn as_socket(&self) -> socket2::SockRef<'_> {
        socket2::SockRef::from(self)
    }
```

Comment (taiki-e) [reply tokio/src/net/udp.rs]: done in fdd5865d9a5cd3b3a86d2b02946ff9dbe26e0afa

**Review by Darksonn — APPROVED** [review]

**Pull request merged**